File: modules/IdTranslator/server/provision.py
from azure.iot.device.aio import ProvisioningDeviceClient
from helpers import compute_derived_symmetric_key

DPS_ENDPOINT = "global.azure-devices-provisioning.net"


class ProvisioningManager:
    def __init__(self, scope_id, group_key, gateway_id, downstream_model_id):
        self._group_key = group_key
        self._scope_id = scope_id
        self._gateway_id = gateway_id
        self._downstream_model_id = downstream_model_id

    async def provision_device(self, device_id):
        provisioning_device_client = ProvisioningDeviceClient.create_from_symmetric_key(
            provisioning_host=DPS_ENDPOINT,
            registration_id=device_id,
            id_scope=self._scope_id,
            symmetric_key=compute_derived_symmetric_key(self._group_key, device_id),
        )

        provisioning_device_client.provisioning_payload = {
            "iotcModelId": self._downstream_model_id,
            "iotcGateway": {"iotcGatewayId": self._gateway_id},
        }
        registration_result = await provisioning_device_client.register()

        if registration_result.status == "assigned":
            return registration_result.registration_state.assigned_hub
        else:
            return None

File: modules/TranslatorNet/Provision.cs
using System;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Azure.Devices.Provisioning.Client;
using Microsoft.Azure.Devices.Provisioning.Client.Transport;
using Microsoft.Azure.Devices.Shared;
using Newtonsoft.Json.Linq;

namespace TranslatorNet
{

    public class Provision
    {
        static readonly string DPS_ENDPOINT = "global.azure-devices-provisioning.net";
        static string _groupKey;
        static string _scopeId;
        static string _gatewayId;
        private static string ComputeDerivedSymmetricKey(string enrollmentKey, string deviceId)
        {
            if (string.IsNullOrWhiteSpace(enrollmentKey))
            {
                return enrollmentKey;
            }

            using var hmac = new HMACSHA256(Convert.FromBase64String(enrollmentKey));
            return Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(deviceId)));
        }
        public static void Init(string scopeId, string groupKey, string gatewayId)
        {
            _scopeId = scopeId;
            _groupKey = groupKey;
            _gatewayId = gatewayId;
        }

        public static async Task<bool> ProvisionDevice(string deviceId, string modelId = null)
        {
            if (_scopeId == null)
            {
                Console.WriteLine("No scope Id has been specified.");
                return false;
            }
            using var security = new SecurityProviderSymmetricKey(
                                       deviceId,
                                       ComputeDerivedSymmetricKey(_groupKey, deviceId),
                                       null);

            using var transportHandler = new ProvisioningTransportHandlerHttp();

            ProvisioningDeviceClient provClient = ProvisioningDeviceClient.Create(
               DPS_ENDPOINT,
                _scopeId,
                security,
                transportHandler);
            Console.WriteLine($"Initialized for registration Id {security.GetRegistrationID()}.");

            Console.WriteLine("Registering with the device provisioning service...");
            JObject provisioningData = new JObject()
            {
                {"iotcGateway", new JObject()
                    {
                        {"iotcGatewayId",_gatewayId}

                    }
                }
            };
            if (modelId != null)
            {
                provisioningData["iotcModelId"] = modelId;
            }
            Console.WriteLine($"Registering with payload '{provisioningData.ToString(Newtonsoft.Json.Formatting.None)}'");
            DeviceRegistrationResult result = await provClient.RegisterAsync(new ProvisioningRegistrationAdditionalData()
            {
                JsonData = provisioningData.ToString(Newtonsoft.Json.Formatting.None)
            });

            Console.WriteLine($"Registration status: {result.Status}.");
            if (result.Status != ProvisioningRegistrationStatusType.Assigned)
            {
                Console.WriteLine($"Registration status did not assign a hub, so exiting this sample.");
                return false;
            }

            Console.WriteLine($"Device {result.DeviceId} registered to {result.AssignedHub}.");
            return true;
        }
    }
}